File: pc/src/valerie/provider/TheMovieDbProvider.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package valerie.provider;

import java.util.List;
import org.jdom.Document;
import valerie.MediaInfo;

/**
 *
 * @author Admin
 */
public class TheMovieDbProvider {

    //////////////////////////////////////////////////
    //////////////////////////////////////////////////
    //////////////////////////////////////////////////
    private String APIKEY = "7bcd34bb47bc65d20a49b6b446a32866";

    private String apiImdbLookup = "http://api.themoviedb.org/2.1/Movie.imdbLookup/<lang>/xml/" + APIKEY + "/<imdbid>";
    private String apiGetInfo = "http://api.themoviedb.org/2.1/Movie.getInfo/<lang>/xml/" + APIKEY + "/<tmdbid>";


    //private String apiImdbLookup = "http://api.themoviedb.org/2.0/Movie.imdbLookup?api_key=" + APIKEY + "&imdb_id=";
    //private String apiSearch = "http://api.themoviedb.org/2.0/Movie.search?api_key=" + APIKEY + "&title=";

    private static final int PLOT_MIN_LEN = 10;
    private static final int IMDBID_MIN_LEN = 3; // tt0

    public boolean getTmdbId(MediaInfo info, org.jdom.Element elem) {
        org.jdom.Element eID = elem.getChild("id");
        if(eID != null && eID.getText() != null && eID.getText().length() > 0) {
            info.TmDbId = eID.getText();
            return true;
        }
        return false;
    }

    public boolean getName(MediaInfo info, org.jdom.Element elem) {
        org.jdom.Element eTitle = elem.getChild("name");
        if(eTitle != null && eTitle.getText() != null && eTitle.getText().length() > 0) {
            info.Title = eTitle.getText();
            return true;
        }
        return false;
    }

    public boolean getOverview(MediaInfo info, org.jdom.Element elem) {
        org.jdom.Element ePlot = elem.getChild("overview");
        if(ePlot != null && ePlot.getText() != null && ePlot.getText().length() > PLOT_MIN_LEN) {
                if(!ePlot.getText().equals("No overview found.")) {
                info.Plot = ePlot.getText().replaceAll("\r\n", " ").replaceAll("\n"," ");
                info.Plot += " [TMDB.ORG]";
                return true;
            }
        }
        return false;
    }

    public boolean getReleased(MediaInfo info, org.jdom.Element elem) {
        org.jdom.Element eYear = elem.getChild("released");
        if(eYear != null && eYear.getText() != null && eYear.getText().length() > 0) {
            String strYear = eYear.getText();
            try {
                info.Year = Integer.parseInt(strYear.substring(0, strYear.indexOf("-")));
                return true;
            }catch(Exception ex) {
            }
        }
        return false;
    }

    public boolean getRuntime(MediaInfo info, org.jdom.Element elem) {
        org.jdom.Element eRuntime = elem.getChild("runtime");
        if(eRuntime != null && eRuntime.getText() != null && eRuntime.getText().length() > 0) {
            info.Runtime = Integer.valueOf(eRuntime.getText());
            return true;
        }
        return false;
    }

    public boolean getRating(MediaInfo info, org.jdom.Element elem) {
        org.jdom.Element eRating = elem.getChild("rating");
        if(eRating != null && eRating.getText() != null && eRating.getText().length() > 0) {
            if(!eRating.getText().equals("0.0")) {
                try {
                    Float fPopularity = Float.valueOf(eRating.getText());
                    info.Popularity = fPopularity.intValue();
                    return true;
                }catch(Exception ex) {
                }
            }
        }
        return false;
    }

    public boolean getGenre(MediaInfo info, org.jdom.Element elem) {
        String genre = "";
        org.jdom.Element eGenres = elem.getChild("categories");
        List<org.jdom.Element> children = eGenres.getChildren("category");
        for(org.jdom.Element eGenre : children) {
            if(eGenre != null && eGenre.getAttributeValue("type").equals("genre"))
                genre += eGenre.getAttributeValue("name") + "|";
        }

        if(genre.length() > 0) {
            info.Genres = genre.substring(0, genre.length() - 1);
            return true;
        }
        return false;
    }

    public boolean getTranslated(org.jdom.Element elem) {
        org.jdom.Element eTranslated = elem.getChild("translated");
        if(eTranslated != null && eTranslated.getText() != null && eTranslated.getText().length() > 0) {
            if(eTranslated.getText().equals("true"))
                return true;
        }
        return false;
    }

    //////////////////////////////
    //////////////////////////////

    public boolean getMovieByImdbID(MediaInfo info) {
        if(info.ImdbId.equals(info.ImdbIdNull))
            return false;

        Document xml = null;
        String url = apiImdbLookup;
        url = url.replaceAll("<imdbid>", String.valueOf(info.ImdbId));
        url = url.replaceAll("<lang>", "en");
        xml = valerie.tools.WebGrabber.getXml(url);

        if (xml == null)
            return false;


        List<org.jdom.Element> moviesList = xml.getRootElement().getChildren("movies");
        if(moviesList.size() <= 0)
            return false;
        List<org.jdom.Element> movieList = moviesList.get(0).getChildren("movie");
        for(org.jdom.Element eMovie : movieList) {
            getTmdbId(info, eMovie);
            //getImdbId(info, eMovie);
            getName(info, eMovie);
            getOverview(info, eMovie);
            getReleased(info, eMovie);
            getRating(info, eMovie);
            getRuntime(info, eMovie);
            getGenre(info, eMovie);

            return true;
        }
        return false;
    }

    public boolean getMovie(MediaInfo info, String lang) {
        if(info.TmDbId.equals(info.TmDbIdNull))
            return false;

        lang = lang.toLowerCase();

        if(lang.equals("en")) // en already parsed using getMovieByImdbID()
            return true;

        Document xml = null;

        String url = apiGetInfo;
        url = url.replaceAll("<tmdbid>", String.valueOf(info.TmDbId));
        url = url.replaceAll("<lang>", lang);
        xml = valerie.tools.WebGrabber.getXml(url);

        if (xml == null)
            return false;

        List<org.jdom.Element> moviesList = xml.getRootElement().getChildren("movies");
        if(moviesList.size() <= 0)
            return false;
        List<org.jdom.Element> movieList = moviesList.get(0).getChildren("movie");
        for(org.jdom.Element eMovie : movieList) {

            if(!getTranslated(eMovie))
                continue;

            //getTmdbId(info, eMovie);
            //getImdbId(info, eMovie);
            getName(info, eMovie);
            getOverview(info, eMovie);
            getReleased(info, eMovie);
            getRating(info, eMovie);
            getRuntime(info, eMovie);
            getGenre(info, eMovie);

            return true;
        }
        return false;
    }

    public boolean getArtById(MediaInfo info) {
        if(info.ImdbId.equals(info.ImdbIdNull))
            return false;

        Document xml = null;
        String url = apiImdbLookup;
        url = url.replaceAll("<imdbid>", String.valueOf(info.ImdbId));
        url = url.replaceAll("<lang>", "en");
        xml = valerie.tools.WebGrabber.getXml(url);

        if (xml == null)
            return false;

        List<org.jdom.Element> moviesList = xml.getRootElement().getChildren("movies");
        if(moviesList.size() <= 0)
            return false;
        List<org.jdom.Element> movieList = moviesList.get(0).getChildren("movie");
        for(org.jdom.Element eMovie : movieList) {
            List<org.jdom.Element> imagesList = eMovie.getChildren("images");
            if(imagesList.size() <= 0)
                continue;
            List<org.jdom.Element> imageList = imagesList.get(0).getChildren("image");
            for(org.jdom.Element eImage : imageList) {
                if (eImage.getAttributeValue("type") != null && eImage.getAttributeValue("type").equals("poster")) {
                    if(info.Poster.length() == 0 && eImage.getAttributeValue("url") != null)
                        info.Poster = eImage.getAttributeValue("url");
                }
                if (eImage.getAttributeValue("type") != null && eImage.getAttributeValue("type").equals("backdrop")) {
                    if(info.Backdrop.length() == 0 && eImage.getAttributeValue("url") != null)
                        info.Backdrop = eImage.getAttributeValue("url");
                }
                
                if(info.Poster.length() > 0 && info.Backdrop.length() > 0)
                    return true;

                continue;
            }
        }
        return false;
    }

}

File: pyvalerie/src/TheMovieDbProvider.py
# -*- coding: utf-8 -*-

import re

import Genres
import WebGrabber

from Plugins.Extensions.ProjectValerie.__common__ import printl2 as printl

#------------------------------------------------------------------------------------------

class TheMovieDbProvider(object):

	APIKEY = u"7bcd34bb47bc65d20a49b6b446a32866"
	PLOT_MIN_LEN = 10
	
	apiImdbLookup = u"http://api.themoviedb.org/2.1/Movie.imdbLookup/<lang>/xml/" + APIKEY + u"/<imdbid>"
	apiGetInfo = u"http://api.themoviedb.org/2.1/Movie.getInfo/<lang>/xml/" + APIKEY + u"/<tmdbid>"
	PLOT_MIN_LEN = 10

	def getElem(self, elem, name):
		if elem.getElementsByTagName(name) != None:
			if len(elem.getElementsByTagName(name)) > 0:
				if elem.getElementsByTagName(name)[0] != None:
					if elem.getElementsByTagName(name)[0].childNodes != None:
						if len(elem.getElementsByTagName(name)[0].childNodes) > 0:
							return elem.getElementsByTagName(name)[0].childNodes[0]
		return None

	def getTmdbId(self, info, elem):
		try:
			eID = self.getElem(elem, "id")
			if eID is not None and eID.data is not None and len(eID.data) > 0:
				info.TmDbId = eID.data
				return info
		except Exception, ex:
			printl("Exception: " + str(ex), self)
		return None

	def getName(self, info, elem):
		try:
			eTitle = self.getElem(elem, "name")
			if eTitle is not None and eTitle.data is not None and len(eTitle.data) > 0:
				info.Title = eTitle.data
				return info
		except Exception, ex:
			printl("Exception: " + str(ex), self)
		return None

	def getOverview(self, info, elem):
		try:
			ePlot = self.getElem(elem, "overview")
			if ePlot is not None and ePlot.data is not None and len(ePlot.data) > self.PLOT_MIN_LEN:
				info.Plot = re.sub(u"\r\n", u" ", ePlot.data)
				info.Plot = re.sub(u"\n", u" ", info.Plot)
				info.Plot += u" [TMDB.ORG]" 
				return info
		except Exception, ex:
			printl("Exception: " + str(ex), self)
		return None

	def getReleased(self, info, elem):
		try:
			eYear = self.getElem(elem, "released")
			if eYear is not None and eYear.data is not None and len(eYear.data) > 0:
				strImdb = eYear.data
				date = strImdb.split(u"-")
				info.Year = int(date[0])
				info.Month = int(date[1])
				info.Day = int(date[2])
				return info
		except Exception, ex:
			printl("Exception: " + str(ex), self)
		return None

	def getRuntime(self, info, elem):
		try:
			eRuntime = self.getElem(elem, "runtime")
			if eRuntime is not None and eRuntime.data is not None and len(eRuntime.data) > 0:
				info.Runtime = int(eRuntime.data.strip())
				return info
		except Exception, ex:
			printl("Exception: " + str(ex), self)
		return None

	def getRating(self, info, elem):
		try:
			eRating = self.getElem(elem, "rating")
			if eRating is not None and eRating.data is not None and len(eRating.data) > 0:
				strRating = eRating.data
				if strRating != u"0.0":
					pos = strRating.find(u".")
					if pos > 0:
						strRating = strRating[0:pos]
					info.Popularity = int(strRating)
					return info
		except Exception, ex:
			printl("Exception: " + str(ex), self)
		return None  

	def getGenre(self, info, elem):
		genre = u""
		try:
			genreList = elem.getElementsByTagName("categories")
			for genreListItem in genreList:
				for eGenre in genreListItem.getElementsByTagName("category"):
					if eGenre.getAttribute("type") == "genre":
						genre += Genres.getGenre(eGenre.getAttribute("name")) + u"|"
			if len(genre) > 0:
				info.Genres = genre[:len(genre) - 1] # Remove the last pipe
				return info
		except Exception, ex:
			printl("Exception: " + str(ex), self)
		return None

	def getTranslated(self, elem):
		try:
			eLang = self.getElem(elem, "translated")
			if eLang is not None and eLang.data is not None and len(eLang.data) > 0:
				if eLang.data == "true":
					return True
		except Exception, ex:
			printl("Exception: " + str(ex), self)
		return False

	###############################################

	def getMovieByImdbID(self, info):
		if info.ImdbId == info.ImdbIdNull:
			printl(" <- None (info.ImdbId == info.ImdbIdNull)", self) 
			return None
		
		url = self.apiImdbLookup
		url = re.sub("<imdbid>", info.ImdbId, url)
		url = re.sub("<lang>",   u"en",	   url)
		xml = WebGrabber.getXml(url)
		
		if xml is None:
			printl(" <- None (xml is None)", self) 
			return None
		
		movieList = xml.getElementsByTagName("movie")
		for eMovie in movieList:
			tmp = self.getTmdbId(info, eMovie)
			if tmp is not None:
				info = tmp
			#tmp = self.getImdbId(info, eMovie)
			#if tmp is not None:
			#	info = tmp
			tmp = self.getName(info, eMovie)
			if tmp is not None:
				info = tmp
			tmp = self.getOverview(info, eMovie)
			if tmp is not None:
				info = tmp
			tmp = self.getReleased(info, eMovie)
			if tmp is not None:
				info = tmp
			tmp = self.getRating(info, eMovie)
			if tmp is not None:
				info = tmp
			tmp = self.getRuntime(info, eMovie)
			if tmp is not None:
				info = tmp
			tmp = self.getGenre(info, eMovie)
			if tmp is not None:
				info = tmp
			
			return info
		
		printl(" <- None (eof)", self) 
		return None

	def getMovie(self, info, lang):
		if info.TmDbId == info.TmDbIdNull:
			printl(" <- None (info.TmDbId == info.TmDbIdNull)", self) 
			return None
		
		lang = lang.lower()
		
		if lang == u"en":
			return info #en already parsed using getSerieByImdbID()
		
		url = self.apiImdbLookup
		url = re.sub("<imdbid>", info.ImdbId, url)
		url = re.sub("<lang>",   lang,		url)
		xml = WebGrabber.getXml(url)
		
		if xml is None:
			printl(" <- None (xml is None)", self) 
			return None
		
		movieList = xml.getElementsByTagName("movie")
		for eMovie in movieList:
			
			if self.getTranslated(eMovie) is False:
				continue
			
			#tmp = self.getTmdbId(info, eMovie)
			#if tmp is not None:
			#	info = tmp
			#tmp = self.getImdbId(info, eMovie)
			#if tmp is not None:
			#	info = tmp
			tmp = self.getName(info, eMovie)
			if tmp is not None:
				info = tmp
			tmp = self.getOverview(info, eMovie)
			if tmp is not None:
				info = tmp
			tmp = self.getReleased(info, eMovie)
			if tmp is not None:
				info = tmp
			tmp = self.getRating(info, eMovie)
			if tmp is not None:
				info = tmp
			tmp = self.getRuntime(info, eMovie)
			if tmp is not None:
				info = tmp
			tmp = self.getGenre(info, eMovie)
			if tmp is not None:
				info = tmp
			
			return info
		
		printl(" <- None (eof)", self)
		return None

	def getArtByImdbId(self, info):
		if info.ImdbId == info.ImdbIdNull:
			printl(" <- None (info.ImdbId == info.ImdbIdNull)", self)
			return None
		
		url = self.apiImdbLookup
		url = re.sub("<imdbid>", info.ImdbId, url)
		url = re.sub("<lang>",   u"en",	   url)
		xml = WebGrabber.getXml(url)
		
		if xml is None:
			printl(" <- None (xml is None)", self)
			return None
		
		movieList = xml.getElementsByTagName("movie")
		for eMovie in movieList:
			for p in eMovie.getElementsByTagName("image"):
				if p.getAttribute("type") == "poster":
					if p.getAttribute("size") == "original" or p.getAttribute("size") == "cover":
						info.Poster = p.getAttribute("url")
				elif p.getAttribute("type") == "backdrop":
					if p.getAttribute("size") == "original":
						info.Backdrop = p.getAttribute("url")
				
				if len(info.Poster) > 0 and len(info.Backdrop) > 0:
					return info
			return info
		
		printl(" <- None (eof)", self)
		return None
